src/state_circuit/state.rs

```diff
@@ -0,0 +1,1609 @@
+
+            // q_stack_not_first is 6 when target is 3, we use 1/6 to normalize the value
+            let inv = F::from_u64(6_u64).invert().unwrap_or(F::zero());
+            let i = Expression::Constant(inv);
+
+            let q_stack_not_first = q_target_cur.clone()
+                * (q_target_cur.clone() - one.clone())
+                * (q_target_cur.clone() - two.clone())
+                * (four.clone() - q_target_cur.clone())
+                * i;
+
+            // q_stack_first is 12, we use 1/12 to normalize the value
+            let inv = F::from_u64(12_u64).invert().unwrap_or(F::zero());
+            let i = Expression::Constant(inv);
+
+            let q_stack_first = q_target_cur.clone()
+                * (two.clone() - q_target_cur.clone())
+                * (three - q_target_cur.clone())
+                * (four.clone() - q_target_cur)
+                * (q_target_next.clone() - one)
+                * (q_target_next.clone() - two)
+                * (four - q_target_next)
+                * i;
+
+            let address_cur = meta.query_advice(address, Rotation::cur());
+            let stack_address_table_zero =
+                meta.query_fixed(stack_address_table_zero, Rotation::cur());
+
+            let q_stack = q_stack_first + q_stack_not_first;
+
+            vec![(q_stack * address_cur, stack_address_table_zero)]
+        });
+
+        // global_counter is in the allowed range:
+        meta.lookup(|meta| {
+            let global_counter = meta.query_advice(global_counter, Rotation::cur());
+            let global_counter_table = meta.query_fixed(global_counter_table, Rotation::cur());
+
+            vec![(global_counter, global_counter_table)]
+        });
+
+        // Memory value (for non-first rows) is in the allowed range.
+        // Memory first row value doesn't need to be checked - it is checked above
+        // where memory init row value has to be 0.
+        meta.lookup(|meta| {
+            let q_target = meta.query_fixed(q_target, Rotation::cur());
+            let one = Expression::Constant(F::one());
+            let three = Expression::Constant(F::from_u64(3));
+            let four = Expression::Constant(F::from_u64(4));
+            let q_memory = q_target.clone()
+                * (q_target.clone() - one)
+                * (three - q_target.clone())
+                * (four - q_target);
+
+            let value = meta.query_advice(value, Rotation::cur());
+            let memory_value_table = meta.query_fixed(memory_value_table, Rotation::cur());
+
+            // q_memory is 4 when target is 2, we use 1/4 to normalize the value
+            let inv = F::from_u64(4_u64).invert().unwrap_or(F::zero());
+            let i = Expression::Constant(inv);
+
+            vec![(q_memory * i * value, memory_value_table)]
+        });
+
+        // padding is in the allowed range:
+        meta.lookup(|meta| {
+            let padding = meta.query_advice(padding, Rotation::cur());
```

Comment:
also need to check q_target is in the allowed range ?

Reply:
No need as the fixed column is the same for all proofs (the verifying keys contain the commitment of fixed columns). So, for the proof it always holds:
- the first MEMORY_ROWS_MAX rows have q_target = 2 (except for the first one which has q_target = 1)
- the first STACK_ROWS_MAX rows have q_target = 3 (except for the first one which has q_target = 1)
- the first STORAGE_ROWS_MAX rows have q_target = 4 (except for the first one which has q_target = 1) (this is yet to be implemented)

Reply:
it should be act as constant variables & polynominal of plonk

Reply:
Not sure what you mean - do you think something needs to be changed?

Reply:
I mean fixed column 's meaning from backend plonk proof system side, that is OK to me now!

Reply:
Ok, cool!